packages/plugins/typescript/compatibility/src/selection-set-to-types.ts

```diff
@@ -77,13 +70,31 @@ export function selectionSetToTypes(
           if (isUnionType(parentType) || isInterfaceType(parentType)) {
             inlineFragmentValue = `DiscriminateUnion<RequireField<${stack}, '__typename'>, { __typename: '${typeCondition}' }>`;
           } else {
-            inlineFragmentValue = `{ __typename: '${typeCondition}' } & Pick<${stack}, ${selection.selectionSet.selections
-              .map(subSelection => (subSelection.kind === Kind.FIELD ? `'${subSelection.name.value}'` : null))
-              .filter(a => a)
-              .join(' | ')}>`;
+            let encounteredNestedInlineFragment = false;
+            const subSelections = selection.selectionSet.selections
+              .map(subSelection => {
+                switch (subSelection.kind) {
+                  case Kind.FIELD:
+                    return `'${subSelection.name.value}'`;
+                  case Kind.FRAGMENT_SPREAD:
+                    return `keyof ${baseVisitor.convertName(subSelection.name.value, { suffix: 'Fragment' })}`;
+                  case Kind.INLINE_FRAGMENT:
+                    encounteredNestedInlineFragment = true;
+                    return null;
```

Comment:
Nested inline fragments cannot be effectively handled because their name cannot be resolved until the next call of `handleTypeNameDuplicates`, which requires recursion into `selectionSetToTypes`. This is an obscure use case, so we just throw.

I think a warning would be better, but I haven't seen infrastructure for emitting warnings. If so, please let me know and we can change this.